lib/reservation_manager.rb

```diff
@@ -0,0 +1,191 @@
+    def list_reservations(date:)
+      date = Hotel::DateRange.validate_date(date)
+      reservations = self.reservations.select do |reservation|
+        date >= reservation.check_in_date && date < reservation.check_out_date
+      end
+      return reservations
+    end
+
+    # I can get the total cost for a given reservation
+    def total_cost(reservation_id:)
+      reservation = @reservations.find { |current_reservation| current_reservation.reservation_id == reservation_id }
+      num_nights = reservation.check_out_date - reservation.check_in_date
+      return num_nights * Hotel::Room.cost
+    end
+
+    # I can view a list of rooms that are not reserved for a given date range
+    def find_available_rooms(check_in_date:, check_out_date:)
+      Hotel::DateRange.new(check_in_date, check_out_date)
+      na_reservations = Hotel::DateRange.get_na_objects(@reservations, check_in_date, check_out_date)
+      na_room_ids_reservations = na_reservations.map { |reservation| reservation.room_id }
+
+      # Given a specific date, and that a room is set aside in a hotel block for that specific date, I cannot reserve or create a block for that specific room for that specific date
+      na_blocks = Hotel::DateRange.get_na_objects(@blocks, check_in_date, check_out_date)
+      na_room_ids_blocks = []
+      na_blocks.each do |block|
+        block.room_ids.each do |i|
+          na_room_ids_blocks << i
+        end
+      end
+
+      na_room_ids = (na_room_ids_blocks + na_room_ids_reservations).uniq
+      return available_rooms = @rooms.reject do |room|
+               na_room_ids.include?(room.room_id)
+             end
+
+      # return available_rooms.map { |room| room.room_id }
+    end
+
+    # I can create a Hotel Block if I give a date range, collection of rooms, and a discounted room rate
+    def create_block(room_ids:, check_in_date:, check_out_date:, discount_rate:)
+      room_ids.each { |room_id| self.class.validate_id(room_id) }
+      Hotel::DateRange.new(check_in_date, check_out_date)
+      available_rooms = find_available_rooms(check_in_date: check_in_date, check_out_date: check_out_date)
+      available_room_ids = available_rooms.map { |room| room.room_id }
+
+      # I want an exception raised if I try to create a Hotel Block and at least one of the rooms is unavailable for the given date range
+      room_ids.each do |room_id|
+        raise ArgumentError, "Room #{room_id} is not available" if available_room_ids.include?(room_id) == false
+      end
+      new_block = Hotel::Block.new(
+        block_id: @blocks.length + 1,
+        room_ids: room_ids,
+        check_in_date: check_in_date,
+        check_out_date: check_out_date,
+        discount_rate: discount_rate,
+      )
+      @blocks << new_block
+      store_blocks_in_csv
+      return new_block
+    end
+
+    # I can check whether a given block has any rooms available
+    def check_available_rooms_in_blocks(block_id:)
```

Comment:
This method doesn't exactly check available rooms, but rather lists available rooms in a block.

Reply:
is this wrong?